# Analytic Hierarchy Process Case study #1: Choosing the next CEO of the company

This is the first Analytic Hierarchy Process case study which has been done to determine who is supposed to be the next contender of the company as a CEO.

The contenders are to be chosen between three individuals, Tom, Dick and Harry

Case study scenario taken from Wikipedia with analytics done using the help of Python.

github = user/clutchkingasiimov

# Decision Scenario

The company, founded in 1960, makes specialized industrial equipment. Its future success will depend on maintaining the strength of its older product lines and on generating a constant flow of new ones. The company's founder is retiring soon, and a consulting firm has developed a detailed plan for continuing its success in his absence. The plan will take five years to implement, and will replace the founder's highly subjective "seat of the pants" style with a more carefully thought out way of doing business.

The board of directors needs to choose someone to lead the company through the change and upheaval that implementing the consultant's plan will involve. In doing this work, the new leader will be required to make many unpopular decisions and take many unpopular actions. He or she will be expected to “clear the air” by stepping aside after the plan is fully implemented.

Six months ago, the board said:

After much thought and discussion, we have identified four criteria to be used in choosing the person to guide us through the upcoming period of change: experience, education, charisma and age. Experience is important because the job requires skills and knowledge that can only be developed through practical application. And though our beloved founder was a self-made man who didn’t finish high school, the times demand that our new leader have an appropriate university education. Since the new leader will have to keep us all motivated during a difficult period of change, we prefer someone with an active, charismatic leadership style. Finally, the new leader's Age is important because he or she will need to have an appropriate career path after stepping down five years from now. — Board of directors, letter to employees and shareholders

Last week, they said:

After an extensive search, we have selected three candidates for this very challenging position. All are presently executives with the company. Choosing among them will be difficult, but we plan to announce our decision shortly. — Board of directors, followup letter to employees and shareholders


# Decision Hierarchy

### The AHP hierarchy for this decision is as follows

##### Goal: Choose the Most Suitable Leader
##### Criteria: Experience, Education, Charisma, Age
##### Alternatives: Tom, Dick, Harry

As the decision makers continue with the AHP, they will determine priorities for the candidates with respect to each of the decision criteria, and priorities for each of the criteria with respect to their importance in reaching the goal.

The priorities will then be combined throughout the hierarchy to give an overall priority for each candidate. The candidate with the highest priority will be the most suitable Alternative, and the ratios of the candidates' priorities will indicate their relative strengths with respect to the Goal.

# Importing Dependencies needed for AHP

In [119]:
import numpy as np
import scipy 
from numpy import linalg

# Step 1 : Pairwise Comparisons

The priorities will be derived from a series of measurements: pairwise comparisons involving all the nodes.

The nodes at each level will be compared, two by two, with respect to their contribution to the nodes above them. The results of these comparisons will be entered into a matrix which is processed mathematically to derive the priorities for all the nodes on the level.

The comparisons can be made in any sequence, but in this example we will begin by comparing the Alternatives with respect to their strengths in meeting each of the Criteria. Then we'll compare the Criteria with respect to their importance to reaching the Goal.

Since there are three alternatives (Tom, Dick, and Harry) and we need to compare each one to each of the others, the decision makers (the Board) will make three pairwise comparisons with respect to each Criterion: Tom vs. Dick, Tom vs. Harry, and Dick vs. Harry. For each comparison, the Board will first judge which member of the pair is weaker with respect to the Criterion under consideration. Then they will assign a relative weight to the other candidate.

### Alternatives vs Criteria no.1
###### Experience

Using their knowledge of the work the leaders will be required to do, the board needs to evaluate the candidates' strengths with respect to experience. Though they have good information about each candidate's work history, there is no such thing as an objective scale for measuring "experience." Thanks to the AHP, the Board will be able to develop a scale, applying only to this one case, that measures the candidates' relative strengths with respect to experience.


###### Alternatives compared with respect to Experience
* Tom [0], Dick [4] (Dick is favoured more)
* Tom = [4], Harry = [1] (Tom is favoured more)
* Dick = [9], Harry = [1] (Dick is favoured more)

The next step is to transfer the weights to a matrix, using a method unique to the AHP. For each pairwise comparison, the number representing the greater weight is transferred to the box of the corresponding color; the reciprocal of that number is put into the box of the color corresponding to the smaller number:

In [120]:
#First step is to construct a pairwise matrix of the scales
#they have been assigned
pairwise_matrix = np.array([[1, 1/4, 4],[4, 1, 9], [1/4, 1/9, 1]])
pairwise_matrix

array([[1.        , 0.25      , 4.        ],
       [4.        , 1.        , 9.        ],
       [0.25      , 0.11111111, 1.        ]])

In [138]:
#Next, find the column total of each row in the matrix and then divide
#each row with its corresponding total to obtain a normalized matrix
column_sum = np.sum(pairwise_matrix, axis=0)
column_sum

array([ 5.25      ,  1.36111111, 14.        ])

In [172]:
#Extract specific elements of the column and then divide them
#to get the normalized matrix
indices1 = [0, 3, 6]
indices2 = [1, 4, 7]
indices3 = [2, 5, 8]

column1 = np.take(pairwise_matrix, indices1)
column2 = np.take(pairwise_matrix, indices2)
column3 = np.take(pairwise_matrix, indices3)
print("Column 1 \n", column1)
print("Column 2\n", column2)
print("Column 3 \n",column3)

Column 1 
 [1.   4.   0.25]
Column 2
 [0.25       1.         0.11111111]
Column 3 
 [4. 9. 1.]


In [198]:
#Divide each row element with its row total
norm_matrix = np.array([[column1 / 5.25], [column2 / 1.36111111], [column3 / 14]])
norm_matrix = norm_matrix.T #Tranposing to switch its R&C back properly
norm_matrix = norm_matrix.reshape(3,3)
norm_matrix

array([[0.19047619, 0.18367347, 0.28571429],
       [0.76190476, 0.73469388, 0.64285714],
       [0.04761905, 0.08163265, 0.07142857]])

In [278]:
#Lastly, find the Principal Right Eigenvector
row_sum = np.sum(norm_matrix, axis=1)
output1 = row_sum.reshape(3,1) * 1/3
print("Principal Eigenvector of Experience \n",output1)
print("\nHighest priority for experience gets awarded to Dick of 0.717")

Principal Eigenvector of Experience 
 [[0.21995465]
 [0.71315193]
 [0.06689342]]

Highest priority for experience gets awarded to Dick of 0.717


##### In terms of Experience, Dick gets awared the highest priority

### Alternatives vs Criteria no.2
###### Education

As they did previously with Experience, the Board now compares pairs of candidates with respect to Education. For each comparison, the Board decides which candidate is the weaker with respect to Education, giving his education a weight of 1. Then, using the AHP Fundamental Scale, they assign a weight to the education of the other candidate. For the sake of our example, we again provide a table summarizing their deliberations:

* Tom [3] Dick [1] (Tom gets favoured)
* Tom [1] Harry [5] (Harry gets favoured)
* Dick [1] Harry [7] (Harry gets favoured)

The next step is to transfer the weights to a matrix, using a method unique to the AHP. For each pairwise comparison, the number representing the greater weight is transferred to the box of the corresponding color; the reciprocal of that number is put into the box of the color corresponding to the smaller number

In [199]:
#Comparison matrix of Education 
pairwise_matrix2 = np.array([[1, 3, 1/5], [1/3, 1, 1/7],[5 , 7,1]])
pairwise_matrix2

array([[1.        , 3.        , 0.2       ],
       [0.33333333, 1.        , 0.14285714],
       [5.        , 7.        , 1.        ]])

In [200]:
#Column total of each column
column_sum2 = np.sum(pairwise_matrix2, axis=0)
column_sum2

array([ 6.33333333, 11.        ,  1.34285714])

In [241]:
#Normalizing the matrix

column11 = np.take(pairwise_matrix2, indices1)
column12 = np.take(pairwise_matrix2, indices2)
column13 = np.take(pairwise_matrix2, indices3)

In [242]:
#Normalized matrix
norm_matrix2 = np.array([[column11 / 6.33333333], [column12 / 11], [column13 / 1.34285714]]).T.reshape(3, 3)
norm_matrix2

array([[0.15789474, 0.27272727, 0.14893617],
       [0.05263158, 0.09090909, 0.10638298],
       [0.78947368, 0.63636364, 0.74468085]])

In [279]:
#Principal eigenvector of norm_matrix2
row_sum2 = np.sum(norm_matrix2, axis=1)
output2 = row_sum2.reshape(3,1) * 1/3
print("Principal Eigenvector of Education\n", output2)
print("\nHighest priority for education goes to Harry with 0.723")

Principal Eigenvector of Education
 [[0.19318606]
 [0.08330788]
 [0.72350606]]

Highest priority for education goes to Harry with 0.723


###### In terms of Education, Harry gets the highest priority

### Alternatives vs Criteria no.3
###### Charisma

* Tom [5] Dick [1] (Tom gets favoured more)
* Tom [9] Harry [1] (Tom get favoured more)
* Dick[4] Harry [1] (Dick gets favoured more)

In [210]:
#Comparison matrix of Charisma
pairwise_matrix3 = np.array([[1,5,9],[1/5,1,4],[1/9,1/4,1]])
pairwise_matrix3

array([[1.        , 5.        , 9.        ],
       [0.2       , 1.        , 4.        ],
       [0.11111111, 0.25      , 1.        ]])

In [211]:
#Column totals
column_sum3 = np.sum(pairwise_matrix3,axis=0)
column_sum3

array([ 1.31111111,  6.25      , 14.        ])

In [224]:
#Normalizing the matrix

column21 = np.take(pairwise_matrix3, indices1)
column22 = np.take(pairwise_matrix3, indices2)
column23 = np.take(pairwise_matrix3, indices3)
column21

array([1.        , 0.2       , 0.11111111])

In [228]:
#Normalized matrix 
norm_matrix3 = np.array([[column21 / 1.31111111],[column22 / 6.25], [column23 / 14.0]]).T.reshape(3, 3)
norm_matrix3

array([[0.76271187, 0.8       , 0.64285714],
       [0.15254237, 0.16      , 0.28571429],
       [0.08474576, 0.04      , 0.07142857]])

In [284]:
#Principal Eigenvector for Charisma
row_sum3 = np.sum(norm_matrix3, axis=1)
output3 = row_sum3.reshape(3,1) * 1/3
print("Principal Eigenvector for Charisma\n", output3)

Principal Eigenvector for Charisma
 [[0.73518967]
 [0.19941889]
 [0.06539144]]


##### In terms of Charisma, Tom gets the highest priority.
Note that even though "charisma" is a highly subjective concept with no imaginable measurement scale, the AHP has allowed the board to measure its relative strength among these three candidates. Also note that with different candidates, or even with different board members, the measurements would likely be different as well. The AHP's measurements apply only to the specific case at hand.

### Alternatives vs Criteria no.4
###### Age

A person's age can be determined with a precision of days or even minutes. But such measurements aren't highly useful in making the decision at hand, since there is more to "Age" in this context than mere chronology.

Also, there are U.S. laws against employment discrimination by age. Anyone using age as a decision factor must be very explicit about their reasons and justifications. In this case, it wouldn't be good if an unsuccessful candidate decided to sue the company on the basis of age discrimination.

Tom, Dick, and Harry are now 50, 60, and 30 years old respectively. Here's a summary of the Board's evaluation of them with respect to Age:

* Tom [1] Dick [3] (Dick is preferred more)
* Tom [5] Harry [1] (Tom is preferred more)
* Dick [9] Harry [1] (Harry is preferred more)

In [236]:
#Comparison matrix of age
pairwise_matrix4 = np.array([[1,1/3,5], [3,1,9],[1/5,1/9,1]])
pairwise_matrix4

array([[1.        , 0.33333333, 5.        ],
       [3.        , 1.        , 9.        ],
       [0.2       , 0.11111111, 1.        ]])

In [239]:
#Column total of the matrix
column_sum4 = np.sum(pairwise_matrix4, axis=0)
column_sum4

array([ 4.2       ,  1.44444444, 15.        ])

In [248]:
#Normalizing the matrix
column31 = np.take(pairwise_matrix4, indices1)
column32 = np.take(pairwise_matrix4, indices2)
column33 = np.take(pairwise_matrix4, indices3)

In [253]:
#Normalized matrix
norm_matrix4 = np.array([[column31 / 4.2],[column32/1.44444444],[column33/15]]).T.reshape(3,3)
norm_matrix4

array([[0.23809524, 0.23076923, 0.33333333],
       [0.71428571, 0.69230769, 0.6       ],
       [0.04761905, 0.07692308, 0.06666667]])

In [287]:
#Principal eigenvector of Age
row_sum4 = np.sum(norm_matrix4, axis=1)
output4 = row_sum4.reshape(3,1) * 1/3
print("Principal Eigenvector of Age\n", output4)
print("\nDick gets the highest priority for Age by 0.668")

Principal Eigenvector of Age
 [[0.26739927]
 [0.66886447]
 [0.06373626]]

Dick gets the highest priority for Age by 0.668


## Criteria vs. the Goal
Now that the decision makers have evaluated the Alternatives (candidates) with respect to their strength in meeting the Criteria, they need to evaluate the Criteria with respect to their importance in reaching the goal.

Once again they do this by a series of pairwise comparisons. As with the Alternatives, this part of the process requires much discussion and debate among the decision makers.

In this case, the Board agrees on these relative weights for the various pairs of Criteria:

* Experience[4] Education[1] 
* Experience[3] Charisma[1]
* Experience[7] Age[1]
* Education[1] Charisma[3]
* Education[3] Age[1]
* Age[1] Charisma[5]

In [260]:
#Comparison matrix of Criteria vs Goal
criteria_matrix = np.array([[1,4,3,7],[1/4,1,1/3,3],[1/3,3,1,5],[1/7,1/3,1/5,1]])
criteria_matrix

array([[1.        , 4.        , 3.        , 7.        ],
       [0.25      , 1.        , 0.33333333, 3.        ],
       [0.33333333, 3.        , 1.        , 5.        ],
       [0.14285714, 0.33333333, 0.2       , 1.        ]])

In [263]:
#Sum of columns
criteria_sum = np.sum(criteria_matrix, axis=0)
criteria_sum

array([ 1.72619048,  8.33333333,  4.53333333, 16.        ])

In [268]:
#Normalizing the matrix
indices11 = [0,4,8,12]
indices12 = [1,5,9,13]
indices13 = [2,6,10,14]
indices14 = [3,7,11,15]

column41 = np.take(criteria_matrix, indices11)
column42 = np.take(criteria_matrix, indices12)
column43 = np.take(criteria_matrix, indices13)
column44 = np.take(criteria_matrix, indices14)

In [273]:
#Normalized matrix 
norm_criteria = np.array([[column41/1.72619048], [column42/8.33333333],[column43/4.53333333],[column44/16]]).T.reshape(4,4)
norm_criteria

array([[0.57931034, 0.48      , 0.66176471, 0.4375    ],
       [0.14482759, 0.12      , 0.07352941, 0.1875    ],
       [0.19310345, 0.36      , 0.22058824, 0.3125    ],
       [0.08275862, 0.04      , 0.04411765, 0.0625    ]])

In [276]:
#Principal Eigenvector of Criteria VS Goal
row_sumcriteria = np.sum(norm_criteria, axis=1)
print("Principal Eigenvector of Criteria VS Goal\n", row_sumcriteria.reshape(4, 1) * 1/4)
print("\nExperience gets the highest priority of 0.539")

Principal Eigenvector of Criteria VS Goal
 [[0.53964376]
 [0.13146425]
 [0.27154792]
 [0.05734407]]

Experience gets the highest priority of 0.539


Note that, in this decision, Experience, the highest ranked Criterion, is about twice as important in reaching the goal as the second-highest ranked Criterion, Charisma. Similarly, Charisma is about twice as important as Education, which in turn is more than twice as important as Age.

## Synthesizing Final Priorities

Now that we know the priorities of the Criteria with respect to the Goal, and the priorities of the Alternatives with respect to the Criteria, we can calculate the priorities of the Alternatives with respect to the Goal. This is a straightforward matter of multiplying and adding, carried out over the whole of the hierarchy

##### Criterion VS Goal for Experience

In [290]:
#Experience Priority for the individuals
experience_criterion = output1 * 0.54
experience_criterion

array([[0.11877551],
       [0.38510204],
       [0.03612245]])

In [293]:
#Education Priority for the individuals
education_criterion = output2 * 0.131
education_criterion

array([[0.02530737],
       [0.01091333],
       [0.09477929]])

In [296]:
#Charisma Priority for the indivuals
charisma_criterion = output3 * 0.271
charisma_criterion

array([[0.1992364 ],
       [0.05404252],
       [0.01772108]])

In [299]:
#Age Priority for the indivuals
age_criterion = output4 * 0.057
age_criterion

array([[0.01524176],
       [0.03812527],
       [0.00363297]])

In [301]:
final_decision_vector = experience_criterion + education_criterion + charisma_criterion + age_criterion
final_decision_vector

array([[0.35856104],
       [0.48818317],
       [0.15225579]])

# Making the Decision

Based on the Board's choice of decision criteria, on their judgments about the relative importance of each, and on their judgments about each candidate with respect to each of the criteria, Dick, with a priority of 0.488, is by far the most suitable candidate. Tom, with a priority of 0.358, is second, and Harry, at 0.152, is third.

The Board should choose Dick as the company's new leader.

Because they have used the AHP, it is easy for them to trace their thinking and to justify the steps along the way to their decision. If they have second thoughts about the final outcome, they can revisit the process and make changes if appropriate. And if they choose to, they can reveal the details of their process to their consultants and confidants, to the candidates, to the shareholders, or to anyone else who might be concerned with the decision.